tool/tsh/app_cloud.go

```diff
@@ -0,0 +1,138 @@
+// Copyright 2023 Gravitational, Inc
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"strings"
+
+	"github.com/gravitational/trace"
+
+	"github.com/gravitational/teleport/lib/client"
+	"github.com/gravitational/teleport/lib/tlsca"
+)
+
+// cloudApp is a dummy interface shared between awsApp, azureApp and gcpApp
+type cloudApp interface {
+	dummyCloudAppMethod()
+}
+
+// cloudAppImpl automatically implements cloudApp when inherited
+type cloudAppImpl struct {
+}
+
+func (_ cloudAppImpl) dummyCloudAppMethod() {
+}
+
+// cloudAppInfo contains common information for given cloudApp
+type cloudAppInfo struct {
+	// matchRouteToApp returns true if given tlsca.RouteToApp is recognized as a given cloud app kind.
+	matchRouteToApp func(tlsca.RouteToApp) bool
+	// newCloudApp creates a new instance of cloudApp
+	newCloudApp func(cf *CLIConf, profile *client.ProfileStatus, appRoute tlsca.RouteToApp) (cloudApp, error)
+	// cloudFriendlyName is a friendly name for given cloud
+	cloudFriendlyName string
+}
+
+func (info *cloudAppInfo) pickActiveCloudApp(cf *CLIConf) (cloudApp, error) {
+	app, needLogin, err := info.pickActiveCloudAppNoRetry(cf)
+	if err != nil {
+		if !needLogin {
+			return nil, trace.Wrap(err)
+		}
+		log.WithError(err).Debugf("Failed to pick an active %v app, attempting to login into app %q", info.cloudFriendlyName, cf.AppName)
+		errLogin := onAppLogin(cf)
```

Comment:
How a user can login into cloud app if the `tsh app proxy` command right now doesn't support any of the following flag: 
```
      --aws-role                 (For AWS CLI access only) Amazon IAM role ARN or role name.
      --azure-identity           (For Azure CLI access only) Azure managed identity name.
      --gcp-service-account      (For GCP CLI access only) GCP service account name.
```

Reply:
Then this approach is inconsistent with `tsh proxy db` / `tsh kube proxy` where a user can specify details for the web app /kube app during the `tsh proxy ... ` call.

Reply:
Adding these flags to `tsh proxy app` shouldn't be too hard, if this is what you suggest?

Reply:
https://github.com/gravitational/teleport/issues/26435